### PR TITLE
DEV-1836 - handle urls in the form 'hifi://<domain>:<port>

### PR DESCRIPTION
**Pull request by roxanneskelly**

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Can someone from High Fidelity approve this PR?

File: libraries/networking/src/AddressManager.cpp [hunk 1]
```diff
@@ -234,14 +234,33 @@ const JSONCallbackParameters& AddressManager::apiCallbackParameters() {
...
+    if (lookupUrl.scheme().isEmpty() && !lookupUrl.path().startsWith("/")) {
+        // 'urls' without schemes are taken as domain names, as opposed to
+        // simply a path portion of a url, so we need to set the scheme
+        lookupUrl.setScheme(URL_SCHEME_HIFI);
+    }
+
+    // it should be noted that url's in the form
+    // somewhere:<port> are not valid, as that
+    // would indicate that the scheme is 'somewhere'
+    // use hifi://somewhere:<port> instead
```
Comment (SimonWalton-HiFi) [inline libraries/networking/src/AddressManager.cpp]: So we don't actually allow machine.highfidelity.com:4321 as a complete domain? It has to have the hifi: prefix?

Comment (roxanneskelly) [reply libraries/networking/src/AddressManager.cpp]: Yeah, 'machine.highfidelity.com:40321' would be a URL with 'machine.highfidelity.com' as the scheme, and 40321 as a relative path.  We could do some coding to support that, of course, but there are edge cases like, what if you had a domain named 'hifi' or 'ftp' or 'http'

Comment (SimonWalton-HiFi) [reply libraries/networking/src/AddressManager.cpp]: I think I'm going to buy the domain ftp!

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/21615/HighFidelity-Beta-PR16273-0589cb3.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/21615/HighFidelity-Quest-Beta-PR16273-0589cb3.apk)

File: libraries/networking/src/AddressManager.cpp [hunk 2]
```diff
@@ -234,14 +234,33 @@ const JSONCallbackParameters& AddressManager::apiCallbackParameters() {
...
+
+    if (lookupUrl.scheme() == URL_SCHEME_HIFI) {
+        if (lookupUrl.host().isEmpty()) {
+            // this was in the form hifi:/somewhere or hifi:somewhere.  Fix it by making it hifi://somewhere
+            const QRegExp HIFI_SCHEME_REGEX = QRegExp(URL_SCHEME_HIFI + ":\\/?", Qt::CaseInsensitive);
```
Comment (SimonWalton-HiFi) [inline libraries/networking/src/AddressManager.cpp]: Could make this `static` for small efficiency gain.

Comment (roxanneskelly) [reply libraries/networking/src/AddressManager.cpp]: will do

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/21615/HighFidelity-Beta-PR16273-0589cb3.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/21615/HighFidelity-Beta-Interface-PR16273-0589cb3.dmg

File: libraries/networking/src/AddressManager.cpp (hunk 1 above)
Comment (SimonWalton-HiFi) [reply libraries/networking/src/AddressManager.cpp]: Currently with Interface, going to 0502-DESKTOP-PC.highfidelity.lan:40110 works. Using the decimal IP also works. We would lose that with this PR?

Comment (roxanneskelly) [reply libraries/networking/src/AddressManager.cpp]: ah will fix.

File: libraries/networking/src/AddressManager.cpp (hunk 2 above)
Comment (SimonWalton-HiFi) [reply libraries/networking/src/AddressManager.cpp]: Also, is the double-backslash necessary? I don't think `/` is a metacharacter.

File: libraries/networking/src/AddressManager.cpp (hunk 1 above)
Comment (SimonWalton-HiFi) [reply libraries/networking/src/AddressManager.cpp]: I think this was the old logic that stuck 'hifi://' in front if it didn't have it. Maybe it needs to be restored.

File: libraries/networking/src/AddressManager.cpp (hunk 2 above)
Comment (roxanneskelly) [reply libraries/networking/src/AddressManager.cpp]: I believe so.

File: libraries/networking/src/AddressManager.cpp (hunk 1 above)
Comment (roxanneskelly) [reply libraries/networking/src/AddressManager.cpp]: unfortunately, the old logic was the code that was messing up the hifi://<domain>:<port> form

Comment (SimonWalton-HiFi) [reply libraries/networking/src/AddressManager.cpp]: Thanks! So the comment immediately above is now out-of-date.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/21621/HighFidelity-Beta-PR16273-9af7b90.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/21621/HighFidelity-Beta-Interface-PR16273-9af7b90.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/21621/HighFidelity-Beta-PR16273-9af7b90.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/21621/HighFidelity-Quest-Beta-PR16273-9af7b90.apk)

**Comment by SimonWalton-HiFi** [comment]
Assuming this works with all the forms in the test plan, as well as host.example.com:4321 and 10.246.56.69:4321, then I think it looks good.

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/21621/HighFidelity-Beta-PR16273-9af7b90.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/21621/HighFidelity-Beta-Interface-PR16273-9af7b90.exe

File: libraries/networking/src/AddressManager.cpp [hunk 3]
```diff
@@ -234,14 +234,38 @@ const JSONCallbackParameters& AddressManager::apiCallbackParameters() {
...
+
+    qCDebug(networking) << "Trying to go to URL" << lookupUrl.toString();
+
+    if (lookupUrl.scheme().isEmpty() && !lookupUrl.path().startsWith("/")) {
+        // 'urls' without schemes are taken as domain names, as opposed to
+        // simply a path portion of a url, so we need to set the scheme
+        lookupUrl.setScheme(URL_SCHEME_HIFI);
+    }
+
+    static const QRegExp PORT_REGEX = QRegExp("\\d{1,5}(\\/.*)?");
```
Comment (SimonWalton-HiFi) [inline libraries/networking/src/AddressManager.cpp]: Still pretty sure you don't need to escape the `/`

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/21640/HighFidelity-Beta-PR16273-fb8a0e3.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/21640/HighFidelity-Beta-Interface-PR16273-fb8a0e3.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/21640/HighFidelity-Beta-PR16273-fb8a0e3.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/21640/HighFidelity-Beta-Interface-PR16273-fb8a0e3.exe

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/21640/HighFidelity-Beta-PR16273-fb8a0e3.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/21640/HighFidelity-Quest-Beta-PR16273-fb8a0e3.apk)

**Pull request merged**